### PR TITLE
apollo assembly optimization 

### PR DESCRIPTION
**Pull request by vdiskg**
## What's the purpose of this PR

1. Assembly Portal, Config Service, Admin Service into a standalone process
2. Assembly ApolloConfigDB, ApolloPortalDB into a single db.

## Which issue(s) this PR fixes:
#4313
#4729

## Brief changelog

1. add table name prefix in the assembly sql
2. add `TablePrefixNamingStrategy` to match the table name prefix

Follow this checklist to help us incorporate your contribution quickly and easily:

- [ ] Read the [Contributing Guide](https://github.com/apolloconfig/apollo/blob/master/CONTRIBUTING.md) before making this pull request.
- [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit tests to verify the code.
- [ ] Run `mvn clean test` to make sure this pull request doesn't break anything.
- [ ] Update the [`CHANGES` log](https://github.com/apolloconfig/apollo/blob/master/CHANGES.md).

### CODE REVIEW

**Comment by nobodyiam** [comment]
Wow! This pr is truly thrilling!

File: apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/AdminServiceAutoConfiguration.java [hunk 1]
```diff
@@ -45,4 +49,19 @@ public FilterRegistrationBean<AdminServiceAuthenticationFilter> adminServiceAuth
...
     return filterRegistrationBean;
   }
+
+  /**
+   * for apollo-assembly
+   */
+  @Order(99)
+  @Profile("assembly")
+  @Configuration
+  static class AdminServiceSecurityConfigurer extends WebSecurityConfigurerAdapter {
```
Comment (nobodyiam) [inline apollo-adminservice/src/main/java/com/ctrip/framework/apollo/adminservice/AdminServiceAutoConfiguration.java:59]: Is the AdminServiceSecurityConfigurer just for assembly stuff? If that's true, maybe it'd make more sense to set it up in a file, something like AssemblyAutoConfiguration?

File: apollo-assembly/src/main/resources/application-github.properties [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#spring.datasource.password=
+
+# Table prefix
+spring.jpa.hibernate.naming.physical-strategy=com.ctrip.framework.apollo.common.jpa.TablePrefixNamingStrategy
+spring.jpa.table-prefix.config-prefix=C_0_
+spring.jpa.table-prefix.portal-prefix=P_0_
+spring.session.jdbc.table-name=P_0_SPRING_SESSION
+
+# H2 datasource
+spring.sql.init.schema-locations=classpath:jpa/@@platform@@/apolloconfigdb.sql, classpath:jpa/@@platform@@/apolloportaldb.sql
```
Comment (nobodyiam) [inline apollo-assembly/src/main/resources/application-github.properties:28]: I observed that four new SQL files have been added: h2/apolloconfigdb.sql, h2/apolloportaldb.sql, mysql/apolloconfigdb.sql, and mysql/apolloportaldb.sql. Were these files generated automatically or edited manually? Additionally, is there a simpler method to manage them?

File: docs/en/deployment/quick-start.md [hunk 3]
```diff
@@ -66,103 +66,99 @@ Quick Start is only for local testing, so generally users do not need to downloa
...
-Just import [sql/apolloportaldb.sql](https://github.com/apolloconfig/apollo-quick-start/blob/master/sql/apolloportaldb.sql) through various MySQL clients.
+# II. Initialization and Startup
+#### Precautions
+1. The Apollo server process needs to use ports 8070, 8080, 8090 respectively, please ensure these three ports are not currently in use.
+2. The `github` in the SPRING_PROFILES_ACTIVE environment variable in the script is a required profile, `auth` is a profile that provides simple authentication for the portal, it can be removed if authentication is not required or other authentication methods are used.
+## 2.1 Use H2 in-memory database, automatic initialization
+No configuration is required, just use the following command to start
+> Note: When using the in-memory database, any operation will be lost after the Apollo process restarts
+```bash
+export SPRING_PROFILES_ACTIVE="github,auth"
```
Comment (nobodyiam) [inline docs/en/deployment/quick-start.md:77]: Is it feasible to enhance the demo.sh script to enable users to initiate the quick-start demonstration efficiently through some configuration settings?

File: apollo-assembly/src/main/resources/application-github.properties (hunk 2 above)
Comment (nobodyiam) [reply apollo-assembly/src/main/resources/application-github.properties:28]: By the way, I observed two new SQL files in the `scripts/sql/assembly` directory. Are these identical to the SQL files located in the `jpa/mysql` folder?

**Review by nobodyiam — COMMENTED** [review]
Now the whole apollo servers could be launched in one process, this is amazing! 
Would you please also update the apollo-development-guide.md?

image

**Comment by nobodyiam** [comment]
It appears there is some incompatibility with database-discovery. When I execute the assembly using the spring profiles `github, auth, database-discovery`, only the apollo-configservice gets registered.

image

File: apollo-build-maven-extensions/pom.xml [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+							<execution>
+								<id>sql-convert</id>
+								<phase>compile</phase>
+								<goals>
+									<goal>java</goal>
+								</goals>
+							</execution>
+						</executions>
+						<configuration>
+							<mainClass>com.ctrip.framework.apollo.maven.extensions.ApolloAssemblySqlConverter</mainClass>
```
Comment (nobodyiam) [inline apollo-build-maven-extensions/pom.xml:57]: This looks interesting. Do the SQL files under `sql/assembly` folder update automatically?

Comment (nobodyiam) [reply apollo-build-maven-extensions/pom.xml:57]: BTW, I met an error trying to create an app:

```
com.google.gson.JsonSyntaxException: com.google.gson.stream.MalformedJsonException: Expected name at line 1 column 3 path $[0].
	at com.google.gson.Gson.fromJson(Gson.java:1241)
	at com.google.gson.Gson.fromJson(Gson.java:1137)
	at com.google.gson.Gson.fromJson(Gson.java:1047)
	at com.google.gson.Gson.fromJson(Gson.java:1014)
	at com.ctrip.framework.apollo.portal.component.config.PortalConfig.organizations(PortalConfig.java:175)
	at com.ctrip.framework.apollo.portal.controller.OrganizationController.loadOrganization(OrganizationController.java:43)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
```

It looks like the `organizations` value is different in the h2 database:

image

While the record in the mysql database is:

image

Comment (vdiskg) [reply apollo-build-maven-extensions/pom.xml]: fixed

**Comment by nobodyiam** [comment]
This is very impressive! I really like the click-and-go experience to start the apollo service with no extra setup.

Below are some comments for your review:
1. Please update the pull request with the latest master branch, as there have been updates to the SQL files.
2. Regarding the quick start guide, my preference is to continue using `demo.sh` for starting/stopping the services. However, we can address this in a separate pull request since it involves modifications to `demo.sh` itself.
3. Some errors were encountered while using the assembly service.
  * Cancel favorite app
image

```
Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "Favorite" not found; SQL statement:
Update Favorite set IsDeleted = true, DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000) where Id = ? [42102-214]
```
  * Create Grayscale Rule
image

```
Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "InstanceConfig" not found; SQL statement:
select b.Id from InstanceConfig a inner join Instance b on b.Id = a.`InstanceId` where a.`ConfigAppId` = ? and a.`ConfigClusterName` = ? and a.`ConfigNamespaceName` = ? and a.`DataChange_LastTime` > ? and b.`AppId` = ? limit ? [42102-214]
```

  * Full release/Abandon grayscale
image

image

```
Caused by: org.h2.jdbc.JdbcSQLSyntaxErrorException: Table "GrayReleaseRule" not found; SQL statement:
Update GrayReleaseRule set IsDeleted = true, DeletedAt = ROUND(UNIX_TIMESTAMP(NOW(4))*1000) where Id = ? [42102-214]
```

**Comment by nobodyiam** [comment]
I'm considering the possibility of connecting our assembly application to two separate databases. This approach could simplify a lot by eliminating the need to rename table names.
It involves establishing two distinct data sources, as outlined in this guide on [Spring Data JPA with Multiple Databases](https://www.baeldung.com/spring-data-jpa-multiple-databases).

**Comment by nobodyiam** [comment]
This pull request appears to be in good shape now. Could you also update the CHANGES.md file? Since this constitutes a major change, it's important to document it there.

**Comment by nobodyiam** [comment]
BTW, #5072 was merged with a schema change, would you please help to update the code base again?

**Review by nobodyiam — COMMENTED** [review]
Please also fix the failed tests.

> 2024-01-26T16:03:31.8898714Z Caused by: java.lang.IllegalStateException: @Order on WebSecurityConfigurers must be unique. Order of 99 was already used on com.ctrip.framework.apollo.adminservice.controller.TestWebSecurityConfig$$EnhancerBySpringCGLIB$$ac5b1e3@469959b6, so it cannot be used on com.ctrip.framework.apollo.adminservice.AdminServiceAssemblyConfiguration$AdminServiceSecurityConfigurer$$EnhancerBySpringCGLIB$$8154145d@270147e3 too.

**Review by nobodyiam — APPROVED** [review]
LGTM

**Pull request merged**